### ISSUE TITLE
Memory pressure due to large image cause crash

### ISSUE THREAD

**Issue by ankittlp** [SDWebImage/SDWebImage#2781]
### New Issue Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios 
 Platform Version        | e.g. 12.0 / 10.14.0 / 12.0 / 5.0
 SDWebImage Version      | e.g. 5.0.0 / 4.4.0
 Integration Method      | e.g.  cocoapods 
 Xcode Version           | e.g. Xcode 10
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

I see there is spike in memory when my image view in table cell tries to dowload the image and show, it cause to crash my app. I have tried to transform image but before even getting to transform image i see a spike im memory of 1.2GB and app crashing.

**Comment by dreampiggy** [comment]
Transforming can not solve the first memory pick during decoding. It just `produce another small image when we have a large image`. So the memory peak exists.

Try to reduce your image resolution side, or use the non-alpha channel image. Because the RAM usage for bitmap image is just:

```
RAM = pixel width * pixel height * bytes per pixel (4 for RGBA8888)
```

**Comment by dreampiggy** [comment]
See more #2560

There is a draft implementation , which use decoding thumb decoding API (like [CGImageSourceCreateThumbnailAtIndex](https://developer.apple.com/documentation/imageio/1465099-cgimagesourcecreatethumbnailatin?language=objc), to produce a fixed size of image (instead of original resolution). Which can avoid peak just during decoding but not image transforming.

**Comment by ankittlp** [comment]
Actually, it's an apple's public API which I can not control. Thus I used transformation delegate and check the size of the image for good (non-crashing limit) size and return the placeholder so as to avoid the crash if size crosses the threshold.

**Comment by dreampiggy** [comment]
Transformer can not solve `The first allocation for full pixels images`. This is most cases you face OOM.

Transformer can only **scale down 10000*10000 pixel into 100*100 and save 10000 times RAM usage, after I already allocate the large buffer**.

This is why that #2560 draft implementation used for.

**Comment by dreampiggy** [comment]
Should be solved by using the thumbnail decoding. Available in v5.5.50. See more in #2922
